### DISCUSSION TITLE
Getting information from channel in channel writer

### DISCUSSION THREAD

**Discussion by skhilliard** [nextgenhealthcare/connect#4514] [Q&A]
What is the recommended approach for getting information back from a channel specified by a channel writer destination?  We have a situation where we delegate work to channels in a somewhat dynamic way to help us accommodate working with many different HIS systems...

![image](https://user-images.githubusercontent.com/23263471/115447511-3477a500-a1de-11eb-955f-dbac9cc78c3c.png)

...the ${mirthChannel} channel could be implemented in any number of ways.  So far, the only way I have found to have access to information that comes from responses in the ${mirthChannel} channel back int the originating channel writer is to put data into the globalMap.  However, that seems like a hack.  Is there a better way to get information (as well as success/failure) from the channel indicated by ${mirthChannel} back to the channel writer invoking it?

Thanks,
Kelly

**Comment by tonygermano** [comment]
You can send any message you want from the downstream channel back upstream as the downstream channel's response. Then you can access it in the response transformer of the calling channel.

**Reply by skhilliard** [reply]
Thanks for the quick reply!  My primary problem was that I had the destination channel configured with no response.  Duh.  Once I configured that correctly, I was able to get access to the response messages as you indicated.

I do have one follow up question...I am now struggling with how to cope with errors that may happen on the channel referenced by channel writer.  For testing, I have a sample JavaScript writer channel I am referencing and forcing errors by doing things like throwing an exception.  In the dashboard, that JavaScript writer channel does show a status of error, but the originating Channel writer that invoked it just shows sent status.

What would be the recommendation for having a channel writer "pick up" on any error condition happening in the referenced channel?  Ideally I think it would be best if I could force the Channel writer to a state of error in that event.

**Reply by skhilliard** [reply]
Oh...another piece of information.  For my test Javascript writer channel....I am returning a response object using the postprocessor script.  Basically, my goal is to realistically simulate a "real world" error condition so I can harden my code to react accordingly.

**Reply by tonygermano** [reply]
It doesn't look like it's possible with a Channel Writer. Here's the source code for it: https://github.com/nextgenhealthcare/connect/blob/development/server/src/com/mirth/connect/connectors/vm/VmDispatcher.java#L91

It looks like it doesn't actually check the value of status of the Response, but just always sets the upstream status to SENT, if the downstream was able to process the message without throwing an exception (for example if the downstream channel is stopped.)

You could probably do that if you route the message manually rather than with a Channel Writer.

Alternatively, you could embed your response message in a new message that contains some of the metadata you care about and then unpack it in the upstream response transformer.
